File: Virtual Machine/Nova.VM/ByteClass.cpp
#include "ByteClass.h"

void ByteClass::Deserialize(BinaryReader& reader)
{
	int methodCount = reader.Read<int>();

	for (int i = 0; i < methodCount; i++)
	{
		std::string methodName = reader.ReadString();
		ByteMethod* method = new ByteMethod(methodName, this);
		method->Deserialize(reader);
		this->Methods.insert(std::make_pair(methodName, method));
	}

	int fieldsCount = reader.Read<int>();

	for (int i = 0; i < fieldsCount; i++)
	{
		std::string fieldName = reader.ReadString();
		ByteField* field = new ByteField(fieldName);
		field->Deserialize(reader);
		this->Fields.insert(std::make_pair(fieldName, field));
	}
}

ByteClass::ByteClass(std::string name)
{
	this->Name = name;
}

File: Nova.Bytecode/IO/ByteClass.cs
using Nova.Bytecode.Enums;
using Nova.Bytecode.Symbols;
using Nova.Utils.IO;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nova.ByteCode.IO
{
    public class ByteClass : IByteElement
    {
        public string Name
        {
            get;
            private set;
        }
        public ContainerType Type
        {
            get;
            private set;
        }
        public List<ByteMethod> Methods
        {
            get;
            set;
        }

        public List<ByteField> Fields
        {
            get;
            set;
        }

        public NovFile NovFile
        {
            get;
            private set;
        }

        public ByteClass(NovFile file, string name, ContainerType type)
        {
            this.NovFile = file;
            this.Name = name;
            this.Type = type;
            this.Methods = new List<ByteMethod>();
            this.Fields = new List<ByteField>();
        }


        public void Serialize(CppBinaryWriter writer)
        {
            writer.Write(Name);

            writer.Write((byte)Type);

            writer.Write(Methods.Count);

            foreach (var method in Methods)
            {
                method.Serialize(writer);
            }

            writer.Write(Fields.Count);

            foreach (var field in Fields)
            {
                field.Serialize(writer);
            }
        }
    }
}
